File: frontend/src/objects/customer/profile/CustomerMetrics.ts
// Business note: frontend mirror of a backend service object; keep names and fields aligned unless this file is explicitly frontend-only.
import type {
  CurrencyCents,
  EntityCount,
  MembershipTier,
} from '@/objects/core/SharedObjects'
import type { Coupon } from '@/objects/customer/profile/Coupon'

export type CustomerMetrics = {
  revokedReviewCount: EntityCount
  membershipTier: MembershipTier
  monthlySpendCents: CurrencyCents
  balanceCents: CurrencyCents
  coupons: Coupon[]
}

File: backend/src/main/scala/services/customer/objects/CustomerMetrics.scala
package services.customer.objects

// Business note: service-owned domain object; mirror it in the frontend only when it is part of the protocol surface.
import system.objects.given

import system.objects.*
import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.*

final case class CustomerMetrics(
    revokedReviewCount: EntityCount,
    membershipTier: MembershipTier,
    monthlySpendCents: CurrencyCents,
    balanceCents: CurrencyCents,
    coupons: List[Coupon],
)
object CustomerMetrics:
  given Encoder[CustomerMetrics] = deriveEncoder
  given Decoder[CustomerMetrics] = deriveDecoder
